# Exploratory Data Analysis of Large-Scale Retail Transaction Data

### Using Univariate, Bivariate and Multivariate Statistical Techniques

## Project Overview

Exploratory Data Analysis (EDA) is a systematic process used to inspect, summarize,
visualize and understand a dataset before applying statistical or machine learning
models.

This project performs detailed exploratory data analysis on a large-scale retail
transaction dataset containing 100,000 records and 18 variables.

The analysis will focus on:

- Understanding the structure and characteristics of the dataset
- Identifying missing, duplicate and invalid values
- Performing data cleaning and preparation
- Conducting univariate, bivariate and multivariate analysis
- Detecting and interpreting outliers
- Measuring correlation between numerical variables
- Analyzing skewness and kurtosis
- Extracting meaningful business insights

### Tools and Libraries

- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn

In [ ]:
#Importing the required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Display configuration
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print("Libraries imported successfully.")

Libraries imported successfully.


### Loading and Intial Exploration of dataset

In [ ]:
df = pd.read_csv("retail_large_dataset.csv")

In [7]:
# Showing the first five rows of the dataset 
df.head()

,customer_id,age,gender,city,state,customer_segment,order_id,order_date,product_category,product_subcategory,product_price,quantity,discount_percentage,final_price,payment_method,shipping_type,delivery_days,return_status
0,C10000,56,Male,Hyderabad,Telangana,New,O50000,2025-05-10,Health,Protein Powder,38658,3,7,107855.82,Debit Card,Standard,5,No
1,C10001,36,Male,Pune,Maharashtra,Premium,O50001,2023-11-27,Electronics,Smartphone,22462,4,20,71878.40,Cash on Delivery,Standard,4,Yes
2,C10002,20,Male,Kolkata,West Bengal,New,O50002,2025-02-08,Fashion,Women Dress,56601,2,27,82637.46,Credit Card,Standard,5,No
3,C10003,39,Female,Jaipur,Rajasthan,New,O50003,2024-04-19,Beauty,Perfume,26158,3,9,71411.34,UPI,Standard,3,No
4,C10004,20,Male,Chennai,Tamil Nadu,Premium,O50004,2024-10-08,Beauty,Shampoo,31240,1,6,29365.60,UPI,Standard,2,No


In [8]:
# Showing the last five rows of the dataset
df.tail()

,customer_id,age,gender,city,state,customer_segment,order_id,order_date,product_category,product_subcategory,product_price,quantity,discount_percentage,final_price,payment_method,shipping_type,delivery_days,return_status
99995,C109995,29,Female,Pune,Maharashtra,Regular,O149995,2024-06-11,Health,Supplements,49293,1,11,43870.77,Debit Card,Standard,9,No
99996,C109996,28,Male,Mumbai,Maharashtra,Premium,O149996,2024-04-27,Fashion,Men T-Shirt,5722,2,22,8926.32,Cash on Delivery,Express,1,No
99997,C109997,44,Female,Jaipur,Rajasthan,Premium,O149997,2024-04-14,Grocery,Snacks,6311,4,11,22467.16,Cash on Delivery,Express,9,No
99998,C109998,45,Male,Jaipur,Rajasthan,Premium,O149998,2024-09-08,Beauty,Shampoo,46888,1,28,33759.36,Debit Card,Standard,1,No
99999,C109999,19,Male,Kolkata,West Bengal,New,O149999,2025-11-15,Health,Supplements,37361,3,3,108720.51,Credit Card,Standard,2,No


In [ ]:
# Displaying the number of rows and columns explicitly
print(f"Number of rows    : {df.shape[0]:,}")
print(f"Number of columns : {df.shape[1]}")

Number of rows    : 100,000
Number of columns : 18


In [12]:
# Showcasing column names in a reader friendly format
for i, column in enumerate(df.columns, start=1):
    print(f"{i}. {column}")

1. customer_id
2. age
3. gender
4. city
5. state
6. customer_segment
7. order_id
8. order_date
9. product_category
10. product_subcategory
11. product_price
12. quantity
13. discount_percentage
14. final_price
15. payment_method
16. shipping_type
17. delivery_days
18. return_status


In [ ]:
# Checking for essential information about dataset such as dtypes , non-null values , memory usage etc.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   customer_id          100000 non-null  object 
 1   age                  100000 non-null  int64  
 2   gender               100000 non-null  object 
 3   city                 100000 non-null  object 
 4   state                100000 non-null  object 
 5   customer_segment     100000 non-null  object 
 6   order_id             100000 non-null  object 
 7   order_date           100000 non-null  object 
 8   product_category     100000 non-null  object 
 9   product_subcategory  100000 non-null  object 
 10  product_price        100000 non-null  int64  
 11  quantity             100000 non-null  int64  
 12  discount_percentage  100000 non-null  int64  
 13  final_price          100000 non-null  float64
 14  payment_method       100000 non-null  object 
 15  shipping_type     

In [ ]:
# Counting the number of columns based on their data types
df.dtypes.value_counts()

object     12
int64       5
float64     1
Name: count, dtype: int64

In [18]:
# Summarizing the dataset to get a statistical overview of numerical columns 
df.describe()

,age,product_price,quantity,discount_percentage,final_price,delivery_days
count,100000.00,100000.00,100000.00,100000.00,100000.00,100000.00
mean,41.56,30164.53,2.50,15.01,64165.72,5.51
std,13.85,17139.68,1.12,8.95,49969.28,2.87
min,18.00,500.00,1.00,0.00,357.00,1.00
25%,30.00,15375.00,1.00,7.00,24705.00,3.00
50%,42.00,30148.00,3.00,15.00,49752.39,5.00
75%,54.00,44957.00,4.00,23.00,94046.07,8.00
max,65.00,59998.00,4.00,30.00,239980.00,10.00


In [ ]:
# Summarizing the categorical columns to get statistical overview 
df.describe(include='object')

,customer_id,gender,city,state,customer_segment,order_id,order_date,product_category,product_subcategory,payment_method,shipping_type,return_status
count,100000,100000,100000,100000,100000,100000,100000,100000,100000,100000,100000,100000
unique,100000,2,10,9,3,100000,1095,6,21,4,2,2
top,C10000,Male,Hyderabad,Maharashtra,New,O50000,2023-04-12,Health,Protein Powder,Credit Card,Express,No
freq,1,50111,10150,19952,33488,1,124,16858,5668,25307,50008,85189


In [20]:
# Checking for number of unique values in each column
df.nunique()

customer_id            100000
age                        48
gender                      2
city                       10
state                       9
customer_segment            3
order_id               100000
order_date               1095
product_category            6
product_subcategory        21
product_price           48449
quantity                    4
discount_percentage        31
final_price             97092
payment_method              4
shipping_type               2
delivery_days              10
return_status               2
dtype: int64

In [28]:
categorical_columns = [
    'gender',
    'city',
    'state',
    'customer_segment',
    'product_category',
    'product_subcategory',
    'payment_method',
    'shipping_type',
    'return_status'
]

for column in categorical_columns:
    print(f"\n{column}")
    print(df[column].unique())


gender
['Male' 'Female']

city
['Hyderabad' 'Pune' 'Kolkata' 'Jaipur' 'Chennai' 'Delhi' 'Ahmedabad'
 'Noida' 'Mumbai' 'Bangalore']

state
['Telangana' 'Maharashtra' 'West Bengal' 'Rajasthan' 'Tamil Nadu' 'Delhi'
 'Gujarat' 'Uttar Pradesh' 'Karnataka']

customer_segment
['New' 'Premium' 'Regular']

product_category
['Health' 'Electronics' 'Fashion' 'Beauty' 'Grocery' 'Home & Kitchen']

product_subcategory
['Protein Powder' 'Smartphone' 'Women Dress' 'Perfume' 'Shampoo' 'Pulses'
 'Headphones' 'Vitamins' 'Blender' 'Supplements' 'Snacks' 'Shoes'
 'Organic Rice' 'Laptop' 'Face Cream' 'Vacuum Cleaner' 'Men T-Shirt'
 'Jeans' 'Tablet' 'Cookware' 'Microwave']

payment_method
['Debit Card' 'Cash on Delivery' 'Credit Card' 'UPI']

shipping_type
['Standard' 'Express']

return_status
['No' 'Yes']


In [22]:
# Checking the date column explicitly to see if it is in the correct format
df['order_date'].head()

0    2025-05-10
1    2023-11-27
2    2025-02-08
3    2024-04-19
4    2024-10-08
Name: order_date, dtype: object

## Column Classification

### Identifier Variables
- customer_id
- order_id

### Numerical Variables
- age
- product_price
- quantity
- discount_percentage
- final_price
- delivery_days

### Categorical Variables
- gender
- city
- state
- customer_segment
- product_category
- product_subcategory
- payment_method
- shipping_type
- return_status

### Date Variable
- order_date

## Initial Observations

Based on the initial inspection of the dataset:

1. The dataset contains 100,000 transaction records and 18 variables.
2. The dataset includes identifier, numerical, categorical and date variables.
3. Customer and order identifiers should not be treated as conventional numerical features.
4. The numerical variables include age, product price, quantity, discount percentage,
   final price and delivery days.
5. The dataset contains categorical information related to customers, products,
   payment methods, shipping and returns.
6. The order date provides an opportunity for time-based analysis.
7. Further investigation is required to identify missing, duplicate, inconsistent
   and invalid values before performing detailed visualization.

# Data Quality Assessment

## 1. Missing Value Analysis

The dataset was checked for missing values across all 18 columns. The initial
inspection showed that every column contains 100,000 non-null records. A final
aggregate check is performed below to confirm whether any missing values are
present.

In [27]:
# Checking total missing values in the dataset
total_missing = df.isnull().sum().sum()

print(f"Total missing values in the dataset: {total_missing}")

Total missing values in the dataset: 0


### Observation

No missing values were detected in the dataset. Therefore, no missing-value
imputation or row removal is required at this stage.

## 2. Duplicate Record Analysis

Duplicate records can cause transactions to be counted more than once and may
distort statistical summaries and visualizations. Therefore, the dataset is
checked for completely duplicated rows.

In [29]:
# Checking for completely duplicated rows
duplicate_count = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_count}")

Duplicate rows: 0


### Observation

No completely duplicated records were found in the dataset. Therefore, no rows
need to be removed on the basis of full-row duplication.

## 3 Categorical Value Validation

In [31]:
# Checking the frequency of values in each categorical column
for column in categorical_columns:
    print(df[column].value_counts(dropna=False))

gender
Male      50111
Female    49889
Name: count, dtype: int64
city
Hyderabad    10150
Ahmedabad    10057
Bangalore    10047
Mumbai       10015
Chennai       9993
Noida         9964
Delhi         9962
Kolkata       9955
Pune          9937
Jaipur        9920
Name: count, dtype: int64
state
Maharashtra      19952
Telangana        10150
Gujarat          10057
Karnataka        10047
Tamil Nadu        9993
Uttar Pradesh     9964
Delhi             9962
West Bengal       9955
Rajasthan         9920
Name: count, dtype: int64
customer_segment
New        33488
Regular    33361
Premium    33151
Name: count, dtype: int64
product_category
Health            16858
Home & Kitchen    16848
Fashion           16613
Grocery           16592
Electronics       16576
Beauty            16513
Name: count, dtype: int64
product_subcategory
Protein Powder    5668
Supplements       5657
Shampoo           5578
Organic Rice      5567
Snacks            5548
Vitamins          5533
Face Cream        5512
Pulses       

In [32]:
# Checking for leading or trailing whitespace in categorical values
whitespace_issues = {}

for column in categorical_columns:
    count = (
        df[column].astype(str)
        .str.strip()
        .ne(df[column].astype(str))
        .sum()
    )
    whitespace_issues[column] = count

pd.Series(whitespace_issues, name='Whitespace Issues')

gender                 0
city                   0
state                  0
customer_segment       0
product_category       0
product_subcategory    0
payment_method         0
shipping_type          0
return_status          0
Name: Whitespace Issues, dtype: int64

### Observation

The categorical variables contain consistent and expected category labels.
No unexpected categories or obvious formatting inconsistencies were identified
during the initial validation. The whitespace check will be used to confirm that
categorical values do not contain leading or trailing spaces.

No categorical-value cleaning is required based on the current validation.

## 4. Numerical Value Validation

Numerical variables are examined for logically invalid values based on their
expected meaning and the observed structure of the dataset. The validation
focuses on age, product price, quantity, discount percentage, final price and
delivery days.

The objective is to distinguish genuinely invalid values from legitimate
extreme observations, which will be analyzed separately during outlier analysis.

In [33]:
# Defining the numerical columns

numerical_columns = [
    'age',
    'product_price',
    'quantity',
    'discount_percentage',
    'final_price',
    'delivery_days'
]


In [34]:
# Creating a consolidated validation summary for numerical variables

numerical_validation = pd.DataFrame({
    'Column': numerical_columns,
    'Minimum': [df[col].min() for col in numerical_columns],
    'Maximum': [df[col].max() for col in numerical_columns],
    'Zero Count': [(df[col] == 0).sum() for col in numerical_columns],
    'Negative Count': [(df[col] < 0).sum() for col in numerical_columns]
})

numerical_validation

,Column,Minimum,Maximum,Zero Count,Negative Count
0,age,18.00,65.00,0,0
1,product_price,500.00,59998.00,0,0
2,quantity,1.00,4.00,0,0
3,discount_percentage,0.00,30.00,3300,0
4,final_price,357.00,239980.00,0,0
5,delivery_days,1.00,10.00,0,0


In [35]:
# Counting logically invalid values for each numerical variable

invalid_numeric_counts = {
    'Invalid Age (< 18)': (df['age'] < 18).sum(),
    'Invalid Product Price (<= 0)': (df['product_price'] <= 0).sum(),
    'Invalid Quantity (<= 0)': (df['quantity'] <= 0).sum(),
    'Invalid Discount (< 0 or > 100)': (
        (df['discount_percentage'] < 0) |
        (df['discount_percentage'] > 100)
    ).sum(),
    'Invalid Final Price (<= 0)': (df['final_price'] <= 0).sum(),
    'Invalid Delivery Days (<= 0)': (df['delivery_days'] <= 0).sum()
}

pd.Series(invalid_numeric_counts, name='Invalid Records')

Invalid Age (< 18)                 0
Invalid Product Price (<= 0)       0
Invalid Quantity (<= 0)            0
Invalid Discount (< 0 or > 100)    0
Invalid Final Price (<= 0)         0
Invalid Delivery Days (<= 0)       0
Name: Invalid Records, dtype: int64

In [36]:
# Checking whether discounts fall outside the expected 0-30% range
outside_discount_range = df[
    (df['discount_percentage'] < 0) |
    (df['discount_percentage'] > 30)
]

print(f"Records outside the 0-30% discount range: {len(outside_discount_range)}")

Records outside the 0-30% discount range: 0


### Observation

The numerical variables do not contain any logically invalid values based on the
validation rules applied. Age values fall within the observed range of 18–65,
product prices and final prices are positive, quantities range from 1–4, and
delivery times range from 1–10 days.

Discount percentages fall entirely within the 0–30% range, with 3,300 records
having a valid discount of 0%. These records represent transactions where no
discount was applied and therefore do not require removal.

No numerical-value cleaning is required based on the current validation.
Potential extreme observations will be investigated separately during outlier
analysis.

## 5. Date Validation

The `order_date` column is currently stored as an object datatype. Before
converting it to datetime, the values are checked for invalid or unparseable
dates and the overall date range is examined.

In [37]:
# Attempting to parse order_date values as datetime
parsed_dates = pd.to_datetime(df['order_date'], errors='coerce')

invalid_date_count = parsed_dates.isna().sum()

print(f"Invalid date values: {invalid_date_count}")

Invalid date values: 0


In [39]:
# Checking the date range
print(f"Earliest order date: {parsed_dates.min()}")
print(f"Latest order date  : {parsed_dates.max()}")

Earliest order date: 2023-01-01 00:00:00
Latest order date  : 2025-12-30 00:00:00


In [40]:
print(f"Unique order dates: {parsed_dates.nunique()}")

Unique order dates: 1095


### Date Continuity Check

In addition to validating individual date values, the order-date range is
checked for missing calendar dates. This helps determine whether the dataset
contains any gaps in its daily timeline.

In [41]:
# Checking whether any calendar dates are missing from the observed date range

expected_dates = pd.date_range(
    start=parsed_dates.min(),
    end=parsed_dates.max(),
    freq='D'
)

missing_dates = expected_dates.difference(parsed_dates)

print(f"Missing calendar dates: {len(missing_dates)}")

Missing calendar dates: 0


### Observation

All values in `order_date` were successfully parsed, with no invalid or
unparseable dates detected. The dataset contains 1,095 unique order dates
between 1 January 2023 and 30 December 2025.

The date range will later be converted to datetime format and used for
time-based exploratory analysis such as monthly, weekly and daily transaction
patterns.

## 6. Business-Rule Validation

In addition to checking individual fields, related numerical variables are
validated against an expected pricing relationship. The calculated final price
is compared with the recorded `final_price` to identify potentially inconsistent
transactions.

The expected relationship is:

Expected Final Price = Product Price × Quantity × (1 − Discount Percentage / 100)

This validation is exploratory because the dataset documentation does not
specify whether additional charges, taxes, shipping costs or other pricing
adjustments are included in `final_price`.

In [42]:
# Calculating the expected final price using the observed pricing components

expected_final_price = (
    df['product_price']
    * df['quantity']
    * (1 - df['discount_percentage'] / 100)
)

In [43]:
# Measuring the difference between recorded and expected final price

price_difference = df['final_price'] - expected_final_price

price_difference.describe()

count   100000.00
mean         0.00
std          0.00
min         -0.00
25%          0.00
50%          0.00
75%          0.00
max          0.00
dtype: float64

In [44]:
# Checking how many transactions closely match the expected pricing formula

tolerance = 1

matching_records = price_difference.abs() <= tolerance

print(f"Matching records: {matching_records.sum():,}")
print(f"Non-matching records: {(~matching_records).sum():,}")
print(f"Match percentage: {matching_records.mean() * 100:.2f}%")

Matching records: 100,000
Non-matching records: 0
Match percentage: 100.00%


### Observation

The recorded `final_price` is fully consistent with the expected pricing
calculation derived from `product_price`, `quantity` and `discount_percentage`.
All 100,000 transactions match the expected calculation within the defined
tolerance, resulting in a 100.00% consistency rate.

Therefore, no pricing inconsistencies were identified and no corrections to
these monetary variables are required based on this validation.

## 7. Data Quality Summary

The dataset underwent a systematic quality assessment covering missing values,
duplicate records, categorical consistency, numerical validity, date validity
and internal pricing consistency.

The analysis found no missing values, complete duplicate records, categorical
formatting issues or logically invalid numerical values. All order dates were
valid and the observed date range contained no missing calendar dates.

Furthermore, the recorded `final_price` was found to be fully consistent with
the expected price calculated from product price, quantity and discount
percentage.

Based on these checks, the dataset is considered suitable for exploratory
visualization without requiring row deletion or value-level correction.
The `order_date` column will be converted to datetime format during the
data-preparation stage.